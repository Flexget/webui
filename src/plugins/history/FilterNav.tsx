import React, { FC, useEffect } from 'react';
import { Toolbar, Tooltip } from '@material-ui/core';
import { css } from '@emotion/core';
import SelectField from 'common/inputs/formik/SelectField';
import { Spacer } from 'common/styles';
import { useFormikContext } from 'formik';
import { useDebounce } from 'utils/hooks';
import DirectionButton from 'common/inputs/formik/DirectionButton';
import { TaskContainer } from 'core/tasks/hooks';
import { useContainer } from 'unstated-next';
import { GroupByFields, GetHistoryOptions, SortByFields } from './types';

const fieldWrapper = css`
  padding: 0.5rem 2rem 0.5rem;
`;

const textField = css`
  width: 7.5rem;
`;

const selectField = css`
  min-width: 20rem;
`;
const groupByFields = [
  {
    value: GroupByFields.Task,
    label: 'Task',
  },
  {
    value: GroupByFields.Time,
    label: 'Time',
  },
];

const sortByFields = [
  {
    value: SortByFields.Details,
    label: 'Details',
  },
  {
    value: SortByFields.Filename,
    label: 'Filename',
  },
  {
    value: SortByFields.Id,
    label: 'ID',
  },
  {
    value: SortByFields.Task,
    label: 'Task',
  },
  {
    value: SortByFields.Time,
    label: 'Time',
  },
  {
    value: SortByFields.Title,
    label: 'Title',
  },
  {
    value: SortByFields.Url,
    label: 'URL',
  },
];

const FilterNav: FC = () => {
  const { values, submitForm } = useFormikContext<GetHistoryOptions>();

  const debouncedValues = useDebounce(values);
  const { tasks } = useContainer(TaskContainer);

  useEffect(() => {
    submitForm();
  }, [...Object.values(debouncedValues), submitForm]);

  return (
    <Toolbar>
      <div css={fieldWrapper}>
        <SelectField
          css={selectField}
          autoFocus
          id="task"
          label="Task Name"
          name="task"
          options={tasks.map(({ name }) => ({ value: name, label: name }))}
        />
      </div>
      <Spacer />
      <div css={fieldWrapper}>
        <SelectField id="sort" name="sort" label="Sort By" options={sortByFields} css={textField} />
      </div>
      <div>
        <SelectField
          id="grouping"
          label="Group By"
          name="grouping"
          options={groupByFields}
          css={textField}
        />
      </div>
      <Tooltip title="Toggle Direction">
        <DirectionButton color="inherit" name="order" id="order" aria-label="Toggle Direction" />
      </Tooltip>
    </Toolbar>
  );
};

export default FilterNav;
