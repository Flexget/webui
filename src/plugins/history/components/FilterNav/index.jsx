import React, { Component } from 'react';
import PropTypes from 'prop-types';
import IconButton from 'material-ui/IconButton';
import { MenuItem } from 'material-ui/Menu';
import SecondaryNav from 'components/SecondaryNav';
import { Spacer } from 'components/styles';
import { RotatingIcon, PaddedTextField } from './styles';

const ENTER_KEY = 13;
const groupByFields = [
  {
    value: 'task',
    label: 'Task',
  },
  {
    value: 'time',
    label: 'Time',
  },
];

const sortByFields = [
  {
    value: 'details',
    label: 'Details',
  },
  {
    value: 'filename',
    label: 'Filename',
  },
  {
    value: 'id',
    label: 'ID',
  },
  {
    value: 'task',
    label: 'Task',
  },
  {
    value: 'time',
    label: 'Time',
  },
  {
    value: 'title',
    label: 'Title',
  },
  {
    value: 'url',
    label: 'URL',
  },
];

class FilterNav extends Component {
  static propTypes = {
    order: PropTypes.oneOf(['asc', 'desc']).isRequired,
    grouping: PropTypes.oneOf(groupByFields.map(({ value }) => value)).isRequired,
    sort: PropTypes.oneOf(sortByFields.map(({ value }) => value)).isRequired,
    toggleOrder: PropTypes.func.isRequired,
    handleChange: PropTypes.func.isRequired,
  };

  state = {
    task: '',
  }

  handleChange = (event) => {
    this.setState({ task: event.target.value });
  }

  handleKeyPress = (event) => {
    const { handleChange } = this.props;
    if (event.which === ENTER_KEY) {
      handleChange('task')(event);
    }
  }

  render() {
    const {
      order,
      grouping,
      sort,
      toggleOrder,
      handleChange,
    } = this.props;

    const {
      task,
    } = this.state;

    return (
      <SecondaryNav>
        <PaddedTextField
          label="Filter By Task"
          id="task-filter"
          placeholder="Task Name"
          value={task}
          onChange={this.handleChange}
          onKeyPress={this.handleKeyPress}
        />
        <Spacer />
        <PaddedTextField
          id="sort-by"
          select
          label="Sort By"
          value={sort}
          onChange={handleChange('sort')}
        >
          {sortByFields.map(({ value, label }) => (
            <MenuItem value={value} key={value}>{label}</MenuItem>
          ))}
        </PaddedTextField>
        <PaddedTextField
          id="group-by"
          label="Group By"
          select
          value={grouping}
          onChange={handleChange('grouping')}
        >
          {groupByFields.map(({ value, label }) => (
            <MenuItem value={value} key={value}>{label}</MenuItem>
          ))}
        </PaddedTextField>
        <IconButton color="inherit" onClick={toggleOrder}>
          <RotatingIcon rotate={order === 'desc'} className="fa fa-chevron-up" />
        </IconButton>
      </SecondaryNav>
    );
  }
}

export default FilterNav;
