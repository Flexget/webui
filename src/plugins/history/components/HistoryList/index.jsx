import React, { forwardRef } from 'react';
import { connect } from 'react-redux';
import { getGroupedHistory } from 'plugins/history/data/selectors';
import HistoryList from './HistoryList';

export function mapStateToProps({ history }, { grouping }) {
  return {
    history: getGroupedHistory(history, grouping),
    hasMore: !history.totalCount || history.items.length < history.totalCount,
  };
}

const ConnectedHistoryList = connect(mapStateToProps)(HistoryList);
export default forwardRef((props, { listRef, scrollRef }) => (
  <ConnectedHistoryList {...props} ref={listRef} scrollRef={scrollRef} />
));
export { HistoryList };
