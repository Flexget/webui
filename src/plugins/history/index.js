import { createAsyncComponent } from 'utils/loading';
import registry from 'core/registry';
import saga from './state/saga';
import reducer from './state/reducer';

export default () =>
  registry.registerPlugin('history', {
    component: createAsyncComponent(() => import('plugins/history/History')),
    routeDisplayName: 'History',
    routeIcon: 'history',
    reducer,
    saga,
  });
