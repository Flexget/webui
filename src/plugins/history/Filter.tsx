import React, { FC, useMemo } from 'react';
import { useInjectContent } from 'core/layout/AppBar/hooks';
import { Direction } from 'utils/query';
import { GroupByFields, SortByFields } from './types';
import FilterNav from './FilterNav';

interface Props {
  order: Direction;
  grouping: GroupByFields;
  sort: SortByFields;
  toggleOrder: () => void;
  handleChange: (key: string) => (e: any) => void;
}

const Filter: FC<Props> = props => {
  useInjectContent(useMemo(() => <FilterNav {...props} />, [props]));

  return null;
};

export default Filter;
