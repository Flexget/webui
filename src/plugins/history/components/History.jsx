import React, { Component } from 'react';
import { findDOMNode } from 'react-dom';
import PropTypes from 'prop-types';
import HistoryList from 'plugins/history/components/HistoryList';
import FilterNav from 'plugins/history/components/FilterNav';
import { NoPaddingWrapper } from 'components/styles';

export default class HistoryPage extends Component {
  static propTypes = {
    getHistory: PropTypes.func.isRequired,
  };

  scroll = React.createRef()

  list = React.createRef()

  state = {
    grouping: 'time',
    sort: 'time',
    order: 'desc',
  };

  getHistory = page => this.props.getHistory({
    page,
    sort_by: this.state.sort,
    order: this.state.order,
    task: this.state.task,
  });

  restartLoader = () => {
    findDOMNode(this.scroll).scrollIntoView(); // eslint-disable-line react/no-find-dom-node
    this.getHistory(1);
    this.scroll.pageLoaded = 1;
  }

  handleChange = key => event => this.setState({ [key]: event.target.value }, this.restartLoader)

  toggleOrder = () => this.setState(({ order }) => ({
    order: (order === 'asc' ? 'desc' : 'asc'),
  }), this.restartLoader)

  render() {
    const { grouping } = this.state;

    return (
      <NoPaddingWrapper>
        <FilterNav
          handleChange={this.handleChange}
          toggleOrder={this.toggleOrder}
          {...this.state}
        />
        <HistoryList
          grouping={grouping}
          getHistory={this.getHistory}
          ref={{
            listRef: this.list,
            scrollRef: this.scroll,
          }}
        />
      </NoPaddingWrapper>
    );
  }
}
