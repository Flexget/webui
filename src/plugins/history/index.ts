import { createAsyncComponent } from 'utils/loading';
import { registerPlugin } from 'core/plugins/registry';
import HistoryIcon from '@material-ui/icons/History';

export default () =>
  registerPlugin('/history', {
    component: createAsyncComponent(() =>
      import(
        /* webpackChunkName: 'HistoryPlugin' */
        'plugins/history/History'
      ),
    ),
    routeDisplayName: 'History',
    routeIcon: HistoryIcon,
  });
