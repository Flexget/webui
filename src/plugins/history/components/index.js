import { connect } from 'react-redux';
import { request } from 'utils/actions';
import { GET_HISTORY } from 'plugins/history/data/actions';
import History from './History';

function mapDispatchToProps(dispatch) {
  return {
    getHistory: data => dispatch(request(GET_HISTORY, data)),
  };
}

export default connect(null, mapDispatchToProps)(History);
export { History };
