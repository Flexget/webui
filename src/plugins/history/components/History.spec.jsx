import React from 'react';
import { shallow } from 'enzyme';
import { History } from 'plugins/history/components';

describe('plugins/history/components', () => {
  it('renders correctly', () => {
    const tree = shallow(<History
      getHistory={jest.fn()}
    />);
    expect(tree).toMatchSnapshot();
  });
});
