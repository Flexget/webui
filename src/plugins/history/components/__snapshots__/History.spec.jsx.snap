// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/history/components renders correctly 1`] = `ShallowWrapper {}`;
